### PR TITLE
✨ envtest: search the assets index for latest of a release series

### PR DESCRIPTION
**Pull request by cbandy**
This allows configuring the `DownloadBinaryAssetsVersion` field of `envtest.Environment` with only the Major.Minor portions of a Kubernetes API version. In that case, the framework searches the asset index for the latest stable version in that release series.

This matches the behavior of `setup-envtest use {major}.{minor}` without the full complexity of version selectors. The leading `v` is now optional, too. For example:
- `"v1.29"` resolves to [v1.29.4](https://github.com/kubernetes-sigs/controller-tools/blob/50b4feecf68291368219a53ef3645dc0de9a4819/envtest-releases.yaml#L24)
- `"1.27"` resolves to [v1.27.1](https://github.com/kubernetes-sigs/controller-tools/blob/50b4feecf68291368219a53ef3645dc0de9a4819/envtest-releases.yaml#L134)
- `"1.24"` resolves to [v1.24.2](https://github.com/kubernetes-sigs/controller-tools/blob/50b4feecf68291368219a53ef3645dc0de9a4819/envtest-releases.yaml#L222)

Setting the field to Major.Minor.Patch or Major.Minor.Patch-PreRelease behaves the same as before:
- Existing binaries are used without going to the network.
- Prereleases can only be selected by spelling them out, precisely as they are in the asset index.

This is an alternative implementation of #3260. It uses the [k8s.io/apimachinery/pkg/util/version](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/version) package and basic comparisons, rather than an external package with range/constraint matchers.

### CODE REVIEW

File: pkg/envtest/binaries.go [hunk 1]
```diff
@@ -111,6 +110,25 @@ type archive struct {
...
+//  1. the SemVer form of s when it refers to a specific Kubernetes release, or
+//  2. the major and minor portions of s when it refers to a release series, or
+//  3. zero values
+func interpretKubernetesVersion(s string) (exact string, major, minor uint) {
+	if v, err := version.ParseSemantic(s); err == nil {
+		return v.String(), 0, 0
+	}
+
+	// See two parseable components and nothing else.
+	if v, err := version.ParseGeneric(s); err == nil && len(v.Components()) == 2 {
```
Comment (alvaroaleman) [inline pkg/envtest/binaries.go:123]: Can we avoid calling `interpretKubernetesVersion` if `binaryAssetssVersion` is the empty string so we can return the error here and don't ~silently default to the latest version~ fall through to use `envtest--$GOOS-GOOARCH.tar.gz` if we can not parse `binaryAssetsVersion`?

File: pkg/envtest/binaries.go [hunk 2]
```diff
@@ -252,34 +279,49 @@ func downloadBinaryAssetsArchive(ctx context.Context, index *index, version stri
...
+		if len(v.PreRelease()) > 0 {
 			continue
 		}
 
-		parsedVersions = append(parsedVersions, v)
+		// Filter on release series, if any.
+		if lower != nil && v.LessThan(lower) {
+			continue
+		}
+		if upper != nil && !v.LessThan(upper) {
```
Comment (alvaroaleman) [inline pkg/envtest/binaries.go]: Nit: `GreaterThan` might be easier to read than `!LessThan`

File: pkg/envtest/binaries_test.go [hunk 3]
```diff
@@ -28,13 +28,93 @@ import (
...
+					if major != tc.major {
+						t.Errorf("expected major %v for %q, got %v", tc.major, input, major)
+					}
+					if minor != tc.minor {
+						t.Errorf("expected minor %v for %q, got %v", tc.minor, input, minor)
+					}
+					if exact != "" {
+						t.Errorf("expected no canonical version for %q, got %q", input, exact)
+					}
+				}
```
Comment (alvaroaleman) [inline pkg/envtest/binaries_test.go]: Avoid nesting conditionals if possible please:

```suggestion
				if tc.exact {
				    return
				}
				if major != tc.major {
					t.Errorf("expected major %v for %q, got %v", tc.major, input, major)
				}
				if minor != tc.minor {
					t.Errorf("expected minor %v for %q, got %v", tc.minor, input, minor)
				}
				if exact != "" {
					t.Errorf("expected no canonical version for %q, got %q", input, exact)
				}
			}
```

File: pkg/envtest/binaries_test.go [hunk 4]
```diff
@@ -28,13 +28,93 @@ import (
...
 
+func TestInterpretKubernetesVersion(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct {
+		name  string
+		in    []string
+		exact bool
+		major uint
+		minor uint
```
Comment (alvaroaleman) [inline pkg/envtest/binaries_test.go]: please name them such that its clear these are expectations for output, i.E. `expectExact`, `expectMajor` etc or something like that

File: pkg/envtest/binaries_test.go [hunk 5]
```diff
@@ -86,6 +166,25 @@ var _ = Describe("Test download binaries", func() {
 		}
 		Expect(actualFiles).To(ConsistOf("some-file"))
 	})
+
+	It("should download binaries of latest stable version of a release series", func(ctx SpecContext) {
```
Comment (alvaroaleman) [inline pkg/envtest/binaries_test.go:194]: Please add a testcase for a version string that can not be parsed

File: pkg/envtest/binaries_test.go [hunk 6]
```diff
@@ -28,13 +28,93 @@ import (
...
+		{
+			name: "one number is not a version",
+			in: []string{
+				"1", "v1", "v001", "1.", "v1.", "1.x",
+			},
+		},
+		{
+			name:  "two numbers are a release series",
+			major: 0, minor: 1,
+			in: []string{"0.1", "v0.1"},
```
Comment (alvaroaleman) [inline pkg/envtest/binaries_test.go]: These testcases would IMHO be less confusing to read if we ordered the properties 1. name, 2. input 3. expected outputs , ther than 1.name, 2. expected outputs, 3.input

File: pkg/envtest/binaries.go (hunk 2 above)
Comment (cbandy) [reply pkg/envtest/binaries.go]: The idea was to have an inclusive lower bound and an exclusive upper bound, but upon writing the comment, I realized that effectively means `major == major && minor == minor`.

I've done away with the less/greater entirely.

File: pkg/envtest/binaries.go (hunk 1 above)
Comment (cbandy) [reply pkg/envtest/binaries.go:123]: I understand what you're asking: When the user fills out the asset version, we should show them an error if it's not version-shaped. I've left this behavior as-is and added Ginkgo tests for error cases. Do you think those error messages are good enough?

> fall through to use `envtest--$GOOS-GOOARCH.tar.gz`

That doesn't happen, thankfully. When we cannot parse what they gave us, we see if it is in the asset index. This is the same behavior as before this PR.

File: pkg/envtest/binaries.go [hunk 7]
```diff
@@ -124,15 +142,22 @@ func downloadBinaryAssets(ctx context.Context, binaryAssetsDirectory, binaryAsse
...
+		// Look for these specific binaries locally before downloading them from the release index.
+		// Use the canonical form of the version from here on.
+		binaryAssetsVersion = "v" + exact
+	} else if binaryAssetsVersion == "" || major != 0 || minor != 0 {
+		// Select a stable version from the release index before continuing.
 		var err error
 		binaryAssetsIndex, err = getIndex(ctx, binaryAssetsIndexURL)
 		if err != nil {
 			return "", "", "", err
 		}
```
Comment (sbueringer) [inline pkg/envtest/binaries.go:160]: Should we add an else branch here that returns to cover error cases?

File: pkg/envtest/binaries_test.go [hunk 8]
```diff
@@ -28,13 +28,97 @@ import (
...
+
+	testCases := []struct {
+		name   string
+		inputs []string
+
+		expectExact       bool
+		expectSeriesMajor uint
+		expectSeriesMinor uint
+	}{
+		{
```
Comment (sbueringer) [inline pkg/envtest/binaries_test.go:52]: Let's add a test case for the empty string

File: pkg/envtest/binaries.go (hunk 7 above)
Comment (cbandy) [reply pkg/envtest/binaries.go:160]: This line _is_ an early return, so I think you're asking a similar question as here: https://github.com/kubernetes-sigs/controller-runtime/pull/3280#discussion_r2280692113

The behavior before and after this PR is to look in the index for what the user typed when we don't understand how to interpret it. The error comes from not finding it (the nonsensical input) in the asset index.

Comment (cbandy) [reply pkg/envtest/binaries.go:160]: There is a Ginkgo test for strange input here: https://github.com/kubernetes-sigs/controller-runtime/blob/60ef409f0bc46f6bae7c5eb016aa10766d1e58c6/pkg/envtest/binaries_test.go#L197-L200

**Comment by cbandy** [comment]
Failure in a timeout test. Looks unrelated:

```
  [FAILED] Expected an error, got nil
  In [It] at: /home/prow/go/src/sigs.k8s.io/controller-runtime/pkg/internal/testing/process/process_test.go:290
```

/retest

File: pkg/envtest/binaries.go (hunk 7 above)
Comment (sbueringer) [reply pkg/envtest/binaries.go:160]: Sorry my bad. I couldn't place the comment at l.164 because of GitHub UI limitations.

I meant if we should add an else branch in l.164.

I think if someone sets `binaryAssetsVersion` to something unparsable neither the if nor the else if branch are hit.

I would prefer if we handle this error case explicitly in an else branch

Comment (cbandy) [reply pkg/envtest/binaries.go:160]: Done.

**Comment by sbueringer** [comment]
Thx!

/lgtm
/approve
/hold cancel
